src/meshes/t8code_mesh.jl

```diff
@@ -108,6 +111,153 @@ Main mesh constructor for the `T8codeMesh` wrapping around a given t8code
+
+            vertices[16] = tree_node_coordinates[1, N, 1, N, i]
+            vertices[17] = tree_node_coordinates[2, N, 1, N, i]
+            vertices[18] = tree_node_coordinates[3, N, 1, N, i]
+
+            vertices[19] = tree_node_coordinates[1, 1, N, N, i]
+            vertices[20] = tree_node_coordinates[2, 1, N, N, i]
+            vertices[21] = tree_node_coordinates[3, 1, N, N, i]
+
+            vertices[22] = tree_node_coordinates[1, N, N, N, i]
+            vertices[23] = tree_node_coordinates[2, N, N, N, i]
+            vertices[24] = tree_node_coordinates[3, N, N, N, i]
+        end
+
+        t8_cmesh_set_tree_vertices(cmesh, i - 1, vertices, 3)
+    end
+
+    # Connect the coarse mesh elements.
+    for i in 1:length(elemIDs)
+        t8_cmesh_set_join(cmesh, elemIDs[i], neighIDs[i], faces[i], duals[i],
+                          orientations[i])
+    end
+
+    t8_cmesh_commit(cmesh, mpi_comm())
+
+    # Init a new forest with just one element per tree.
+    do_face_ghost = mpi_isparallel()
+    scheme = t8_scheme_new_default_cxx()
+    initial_refinement_level = 0
+    forest = t8_forest_new_uniform(cmesh, scheme, initial_refinement_level, do_face_ghost,
+                                   mpi_comm())
+
+    cum_sum_num_elements_per_tree = cumsum(num_elements_per_tree)
+
+    global_element_id = 0 # zero-based index
+
+    # Compute the offset within the to-be-reconstructed forest. Depends on the
+    # MPI rank resp. first global tree id.
+    if mpi_rank() > 0
+        last_global_tree_id_of_preceding_rank = t8_forest_global_tree_id(forest, 0) - 1
+        global_element_id += cum_sum_num_elements_per_tree[last_global_tree_id_of_preceding_rank + 1]
+    end
+
+    function adapt_callback(forest, local_tree_id, eclass_scheme, local_element_id,
+                            elements, is_family,
+                            user_data)
+
+        # Check if we are already in the next tree in terms of the `global_element_id`.
+        global_tree_id = t8_forest_global_tree_id(forest, local_tree_id)
+        if global_element_id + 1 > cum_sum_num_elements_per_tree[global_tree_id + 1]
+            return 0
+        end
+
+        # Test if we already reached the targeted level.
+        level = t8_element_level(eclass_scheme, elements[1])
+        if level < levels[global_element_id + 1]
+            return 1 # Go one refinement level deeper.
+        end
+
+        # Targeted level is reached.
+        global_element_id += 1
+        return 0
+    end
+
+    # The adapt callback refines the forest according to the `levels` array.
+    # For each tree the callback recursively increases the refinement level
+    # till it matches with the associated section in `levels.
+    forest = adapt(forest, adapt_callback; recursive = true, balance = false,
+                   partition = false, ghost = false, user_data = C_NULL)
+
+    if mpi_isparallel()
+        forest = partition(forest)
+    end
+
```

Comment:
How about a final check that the number of elements reached matches with the given `nelements`?

Reply:
Good idea! Added.